output/write.go

```diff
@@ -0,0 +1,112 @@
+package output
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+// FileInfo returns the filesystem properties atomicWrite wrote
+type FileInfo struct {
+	Mode os.FileMode
+	UID  int
+	GID  int
+}
+
+// GetFileInfo from an open file
+func GetFileInfo(file *os.File) (*FileInfo, error) {
+	stat, err := file.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("failed to stat after writing: %v", err)
+	}
+	filemode := stat.Mode()
+	uid := int(stat.Sys().(*syscall.Stat_t).Uid)
+	gid := int(stat.Sys().(*syscall.Stat_t).Gid)
+
+	return &FileInfo{filemode, uid, gid}, nil
+}
+
+// WriteFileAtomically creates a temporary file, sets perms, writes content, and renames it to filename
+// This sequence ensures the following:
+// 1. Nobody can open the file before we set owner/permissions properly
+// 2. Nobody observes a partially-overwritten secret file.
+// The returned FileInfo may not match the passed in one, especially if chownFiles is false.
+func WriteFileAtomically(dir, filename string, chownFiles bool, fileInfo FileInfo, enforceFilesystem Filesystem, content []byte) (*FileInfo, error) {
+	if err := os.MkdirAll(dir, 0775); err != nil {
+		return nil, fmt.Errorf("making client directory '%s': %v", dir, err)
+	}
+
+	// We can't use ioutil.TempFile because we want to open 0000.
+	buf := make([]byte, 32)
+	_, err := rand.Read(buf)
+	if err != nil {
+		return nil, err
+	}
+	randSuffix := hex.EncodeToString(buf)
+	fullPath := filepath.Join(dir, filename)
+	f, err := os.OpenFile(fullPath+randSuffix, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0000)
+	// Try to remove the file, in event we early-return with an error.
+	defer os.Remove(fullPath + randSuffix)
+	if err != nil {
+		return nil, err
+	}
+
+	if chownFiles {
+		err = f.Chown(fileInfo.UID, fileInfo.GID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Always Chmod after the Chown, so we don't expose secret with the wrong owner.
+	err = f.Chmod(fileInfo.Mode)
+	if err != nil {
+		return nil, err
+	}
+
+	if enforceFilesystem != 0 {
+		good, err := isFilesystem(f, enforceFilesystem)
+		if err != nil {
+			return nil, fmt.Errorf("checking filesystem type: %v", err)
+		}
+		if !good {
+			return nil, fmt.Errorf("unexpected filesystem writing %s", filename)
```

Comment:
nit: would logging the expected filesystem here leak important information?  If not, might be useful for debugging.

Reply:
This is related to issue #54 where it's hard to understand/debug the filesystem checking.  Added a note about this, there, for future improvement